### PR TITLE
Pairs

### PR DESCRIPTION
**Pull request by nojaf**
Hi @safesparrow, I was able to rework the code to have a new version of the infamous `fsix` approach.
I went with the names `Pair` and `Physical` and transform the resolved graph after `mkGraph`.
This has quite the appeal to me as it is more of an implementation detail isolated from the graph resolution.

I wrote some comments on why we do this, for future readers. Please take a look at this and let me know if it makes sense.

### CODE REVIEW

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs [hunk 1]
```diff
@@ -27,12 +27,17 @@ let DiagnosticsLoggerForInput (tcConfig: TcConfig, input: ParsedInput, oldLogger
...
+type SingleResult = State -> FinalFileResult * State
 type Item = File
 
 type PartialResult = TcEnv * TopAttribs * CheckedImplFile option * ModuleOrNamespaceType
 
-let folder (isFinalFold: bool) (state: State) (result: SingleResult) : FinalFileResult * State = result isFinalFold state
+let folder (state: State) (result: SingleResult) : FinalFileResult * State = result state
+
+[<RequireQualifiedAccess>]
+type Dependency =
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:38]: The `Dependency` name seems controversial to me - it represents a node in the graph, right, not a node's dependencies (although node's dependencies are nodes themselves)?

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs [hunk 2]
```diff
@@ -51,7 +56,67 @@ let CheckMultipleInputsInParallel
...
+    // By doing this, we can speed up the graph processing as type checking a signature file is less expensive than its implementation counterpart.
+    //
+    // When we need to actually type-check the implementation file of a Pair, we cannot have the duplicate information of the signature file present in TcEnvFromImpls.
+    // Example `A.fs -> [ A.fsi ]`, because an implementation file always depends on its signature.
+    // Type-checking `A.fs` will add the actual information to TcEnvFromImpls and we do not depend on the Pair<A> for `A.fs` itself.
+    //
+    // In order to deal correctly with the Pair logic, we need to transform the resolved graph to contain the additional pair nodes.
+    // After we have type-checked the graph, we exclude the Pair nodes as they are not actual physical files in our project.
+    //
+    // (1) : A pair is consider the combination of a implementation and signature file. (For example `A.fsi` and matching `A.fs` is Pair<A>)
```
Comment (safesparrow) [inline tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:82]: What does `Dependency.Pair 1` mean then? It probably doesn't mean that we depend on both the .fs and .fsi files?

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs (hunk 1 above)
Comment (nojaf) [reply tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:38]: You are right, I'll change it to `Node`, makes more sense.

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs (hunk 2 above)
Comment (nojaf) [reply tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:82]: `Dependency.Pair 1` means you depend on the information of a signature file with index `1`.
We do need to information to be present in `TcEnvFromSignatures` and `TcEnvFromImpls`.
I'm open to suggestions for the name. It is hard to give it a proper name. 
We basically mean a copy of the processed signature file for the `TcEnvFromImpls`.

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs (hunk 1 above)
Comment (safesparrow) [reply tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:38]: >If Node weren't the most commonly used name, I'd vote for that.
I actually meant that `Node` is very generic so maybe not best, but perhaps it's actually ok 🙂 (in the more generic `GraphProcessing.fs` code I'm using `ProcessedNode` (in a separate module). 
But the bottom line is that `Dependency` could just as well be named `Dependant`, and we should avoid both.

File: tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs (hunk 2 above)
Comment (safesparrow) [reply tests/ParallelTypeCheckingTests/Code/ParallelTypeChecking.fs:82]: Given that it's the `.fsi` file that brings us the signature info in the TcState, could this name focus on what this node adds to the TcState? No concrete name at this point, but `Pair` suggests like it adds both, which is not the case - even if its dependencies do. Instead, the graph itself already presents the information about what dependencies each node has and what they do.

`Node.FakeImplFile` is my poor attempt.

**Review by safesparrow — APPROVED** [review]
LGTM, happy for this to be merged assuming it passes the recently added compilation test.

Thanks @nojaf

**Pull request merged**